src/metabase/measures/api.clj

```diff
@@ -0,0 +1,107 @@
+   ^{:clj-kondo/ignore [:discouraged-namespace]} [metabase.legacy-mbql.schema :as mbql.s]
+   [metabase.lib.core :as lib]
+   [metabase.models.interface :as mi]
+   [metabase.util :as u]
+   [metabase.util.log :as log]
+   [metabase.util.malli :as mu]
+   [metabase.util.malli.schema :as ms]
+   [metabase.xrays.core :as xrays]
+   [toucan2.core :as t2]))
+
+(api.macros/defendpoint :post "/"
+  "Create a new `Measure`."
+  [_route-params
+   _query-params
+   {:keys [name description table_id definition], :as body} :- [:map
+                                                                [:name        ms/NonBlankString]
+                                                                [:table_id    ms/PositiveInt]
+                                                                [:definition  ms/Map]
+                                                                [:description {:optional true} [:maybe :string]]]]
+  (api/create-check :model/Measure body)
+  (let [measure (api/check-500
+                 (first (t2/insert-returning-instances! :model/Measure
+                                                        :table_id    table_id
+                                                        :creator_id  api/*current-user-id*
+                                                        :name        name
+                                                        :description description
+                                                        :definition  definition)))]
+    (events/publish-event! :event/measure-create {:object measure :user-id api/*current-user-id*})
+    (t2/hydrate measure :creator)))
+
+(mu/defn- hydrated-measure [id :- ms/PositiveInt]
+  (-> (api/read-check (t2/select-one :model/Measure :id id))
+      (t2/hydrate :creator)))
+
+(api.macros/defendpoint :get "/:id"
+  "Fetch `Measure` with ID."
+  [{:keys [id]} :- [:map
+                    [:id ms/PositiveInt]]]
+  (hydrated-measure id))
+
+(api.macros/defendpoint :get "/"
+  "Fetch *all* `Measures`."
+  []
+  (as-> (t2/select :model/Measure, :archived false, {:order-by [[:%lower.name :asc]]}) measures
+    (filter mi/can-read? measures)
+    (t2/hydrate measures :creator)))
+
+(defn- write-check-and-update-measure!
+  "Check whether current user has write permissions, then update Measure with values in `body`. Publishes appropriate
+  event and returns updated/hydrated Measure."
+  [id {:keys [revision_message], :as body}]
+  (let [existing   (api/write-check :model/Measure id)
+        clean-body (u/select-keys-when body
+                                       :present #{:description}
+                                       :non-nil #{:archived :definition :name})
+        new-def    (when-let [def (:definition clean-body)]
+                     (cond->> def
+                       (not= :mbql-version/mbql5 (lib/normalized-mbql-version def))
+                       (mbql.normalize/normalize ::mbql.s/MBQLQuery)))
```

Comment:
Segments have to deal with their legacy form; measures are new. This may as well throw a 400 if the definition doesn't validate as MBQL 5 immediately.